programs/stake/src/stake_state.rs

```diff
@@ -757,33 +761,71 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             match self.state()? {
                 StakeState::Stake(meta, mut stake) => {
                     meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                    let split_rent_exempt_reserve = calculate_split_rent_exempt_reserve(
+                        &meta,
+                        self.data_len()? as u64,
+                        split.data_len()? as u64,
+                    );
 
                     // verify enough lamports for rent in new stake with the split
                     if split.lamports()? + lamports < meta.rent_exempt_reserve
-                     // verify enough lamports left in previous stake and not full withdrawal
+                    // verify enough lamports left in previous stake and not full withdrawal
                         || (lamports + meta.rent_exempt_reserve > self.lamports()? && lamports != self.lamports()?)
+                    // verify that the same rent_exempt_reserve is sufficient to make the new
+                    // stake account rent exempt needed because the new uninitialized may be any
+                    // size larger than std::mem::size_of::<StakeState>()
+                        || split_rent_exempt_reserve > meta.rent_exempt_reserve
                     {
                         return Err(InstructionError::InsufficientFunds);
                     }
                     // split the stake, subtract rent_exempt_balance unless
-                    //  the destination account already has those lamports
-                    //  in place.
-                    // this could represent a small loss of staked lamports
-                    //  if the split account starts out with a zero balance
-                    let split_stake = stake.split(
-                        lamports - meta.rent_exempt_reserve.saturating_sub(split.lamports()?),
-                    )?;
+                    // the destination account already has those lamports
+                    // in place.
+                    // this means that the new stake account will have a stake equivalent to
+                    // lamports minus rent_exempt_reserve if it starts out with a zero balance
+                    let (remaining_stake_delta, split_stake_amount) = if lamports
+                        == self.lamports()?
+                    {
+                        // If split amount equals the full source stake, the new split stake must
+                        // equal the same amount, regardless of any current lamport balance in the
+                        // split account. Since split accounts retain the state of their source
+                        // account, this prevents any magic activation of stake by prefunding the
+                        // split account.
+                        (
+                            lamports - meta.rent_exempt_reserve,
+                            lamports - meta.rent_exempt_reserve,
```

Comment:
Also, this must be `lamports - split_rent_exempt_reserve.rent_exempt_reserve`, right?

Reply:
No, definitely not. The amount of stake delegated should not be able to increase by splitting to a smaller-sized account.

Reply:
Hhm. That's because of bad side-effect to the stake distribution for the current epoch? Maybe I'm lacking some precondition.

Reply:
Hmm, it looks like the max stake delta would be ~69 SOL. Is that small enough we don't care?